### DISCUSSION TITLE
Separating CS from lpspi

### DISCUSSION THREAD

**Discussion by ghost** [mciantyre/teensy4-rs#131] [Q&A] [answered]
Currently, `lpspi` requires, or seems to require, a CS pin. Some displays, however, have an extra data/command pin that must be toggled during data transmission. The method for those is:

- Pull CS low manually
- Pull D/C low manually
- Send the command info over the bus
- Pull D/C high
- Send the data over the bus
- Pull CS high, ending the transaction

Can I create an `lpspi4` that doesn't change the state of the CS pin?

Many thanks.

**Comment by mciantyre** [comment, marked as answer]
It sounds like [the `Lpspi::without_pins()` constructor](https://docs.rs/imxrt-hal/0.5.0/imxrt_hal/lpspi/struct.Lpspi.html#method.without_pins) would help. This gives us the driver handle, but it doesn't configure any pins. The caller is responsible for setting up the pins. There's more information [here](https://docs.rs/imxrt-hal/0.5.0/imxrt_hal/lpspi/index.html#chip-selects-cs-for-spi-peripherals).

The BSP lets us write this, but we have to ignore the `board` APIs. e985b7f shows how to rewrite the `blocking_spi_loopback.rs` example to create an LPSPI4 driver without the four required pins.

---

For full transparency: I've never personally tested the "drive chip select with a GPIO" approach. Additionally, there's an LPSPI command queue, and writes block until there's space in the queue. Writes can return early before data is actually sent out. Keep this in mind when changing GPIO states in software. (I don't think there's a `flush()` method in today's HAL, but we could change that if it could help here.)

**Reply by ghost** [reply]
Yes, I'm seeing a lot of (e.g. Adafruit) breakout boards where the IM[0-2] pins aren't wired out and the default is four-wire SPI.  This would definitely be a good edition to support those unfortunately-configured breakouts.

I have some time tonight, so I'm going to play with the manual control of CS/DC a bit. Another question: when I build the SPI from the without_pins constructor, it's not clear to me what type it is (for passing to functions, etc.) Can you let me know what's the optimal passing strategy?

Many, many thanks. This is my first project on Rust, and I'm having a lot of fun, partly because Rust is gorgeous, and partly because your code is clean and well-organized. Thank you for all you do.

**Reply by mciantyre** [reply]
When supplied with the LPSPI**4** peripheral instance, the return type of `without_pins` is a `Lpspi<(), 4>`. We can use that precise type, or a type alias, in functions and APIs. f6d62bc is an extension of the previous demo to show the LPSPI type and how it can be used in a function.
